src/tpm.rs

```diff
@@ -29,30 +29,510 @@ const RETRY: usize = 4;
+    config_pw: String,
+) -> Result<(String, String, String, String, String), KeylimeTpmError> {
+    emulator_warning();
+    tpm_set_password(config_pw)?;
+    tpm_create_ek(Some("sha256".into()))?;
+    tpm_get_pub_ek()?;
+    let ekcert = tpm_read_ekcert_nvram()?;
+    set_tpm_metadata_content("ekcert", &ekcert);
+    tpm_create_aik(
+        self_activate,
+        "rsa".into(),
+        "sha256".into(),
+        "rsassa".into(),
+    )?;
+
+    let ek = get_tpm_metadata_content("ek")?;
+    let aik = get_tpm_metadata_content("aik")?;
+    let ek_tpm = get_tpm_metadata_content("ek_tpm")?;
+    let aik_name = get_tpm_metadata_content("aik_name")?;
+    Ok((ek, ekcert.into(), aik, ek_tpm, aik_name))
+}
+
+/*
+ *  Return:
+ *      ek certificate to prove ek is generated by TPM
+ *      Error if the retrieve process failed
+ *
+ *  Retrieve the certifcate of ek from nvram
+ */
+fn tpm_read_ekcert_nvram() -> Result<String, KeylimeTpmError> {
+    let mut tf = NamedTempFile::new()?;
+    let nvpath = temp_file_get_path(&tf)?;
+    let (ret_out, _) = run("tpm2_nvlist".to_string(), None)?;
+    let ret_out_value: Value = serde_yaml::from_str(&ret_out)?;
+    if ret_out_value["0x1c00002"]["size"].is_null() {
+        return Err(KeylimeTpmError::new_tpm_rust_error(
+            "TPM nvlist failed.",
+        ));
+    }
+
+    let ekcert_size = ret_out_value["0x1c00002"]["size"].clone();
+    let (_, f_out) = run(
+        format!("tpm2_nvread -x 0x1c00002 -s {} -f {}", ekcert_size, nvpath),
+        Some(vec![nvpath]),
+    )?;
+    let ekcert = f_out.get(nvpath).ok_or_else(|| {
+        KeylimeTpmError::new_tpm_rust_error(
+            "failed to read nvram and retrieve ekcert.",
+        )
+    })?;
+    Ok(base64::encode(ekcert))
+}
+
+/*
+ *  Return:
+ *     value from nvram
+ *     Error if the process failed
+ *
+ *  Read the data stored in a Non-Volatile (NV)s index.
+ */
+fn tpm_read_key_nvram() -> Result<String, KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let (ret_out, _) = run(
+        format!(
+            "tpm2_nvread -x 0x1500018 -a 0x40000001 -s {} -P {}",
+            common::BOOTSTRAP_KEY_SIZE,
+            owner_pw
+        ),
+        None,
+    )?;
+
+    if ret_out.len() as i32 == common::BOOTSTRAP_KEY_SIZE {
+        return Ok(ret_out.into());
+    }
+    Ok(String::new())
+}
+
+/*
+ *  Input:
+ *      key to write to nvram
+ *  Return:
+ *      Error if process failed
+ *
+ *  Write data to a Non-Volatile (NV) index.
+ */
+fn tpm_write_key_nvram(key: String) -> Result<(), KeylimeTpmError> {
+    let owner_pw = get_tpm_metadata_content("owner_pw")?;
+    let mut tf = NamedTempFile::new()?;
+    tf.write_all(key.as_bytes())?;
+    let attrs = "ownerread|policywrite|ownerwrite";
+    let tf_path = temp_file_get_path(&tf)?;
+
+    run(format!("tpm2_nvdefine -x 0x1500018 -a 0x40000001 -s {} -t \"{}\" -p {} -P {}", common::BOOTSTRAP_KEY_SIZE, attrs, owner_pw, owner_pw), None)?;
```

Comment:
We might need to implement a check for legacy_tools, if we want to work with rhel 7:

https://github.com/keylime/keylime/blob/master/keylime/tpm2.py#L1031

https://github.com/keylime/keylime/blob/94328329d2dad5d5a659168865d2e73347a48ce4/keylime/tpm2.py#L49-L53

You can see where this is a decision based on the